modules/setting/queue.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package setting
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/queue"
```

Comment:
I don't think package setting should depend on module queue.

Reply:
Ok, it's a shame because queue is otherwise quite well contained depending only on log and could be broken out. I guess so long as the dependency on setting is well contained it could gain it. 

The other option is to stick the setting dependency in service but that means models can't create queues - which I think would be limiting at present.

Reply:
Are the dependencies needed because of the config data types? Perhaps there could be a `queue_config` package that holds only the structs (like the `structs` package). I think `settings` could safely depend on that package as it will not chain any additional weird dependencies.

Reply:
Done

Reply:
There's always this problem with configuration - you don't want configuration to end up depending on everything however, correct configuration depends on the specifics of the modules it's configuring.

In this case the original code would check against the list of valid queue types at configuration time and would do some nice bridging between the ini configuration of our current setting format and the more POJO like config I'd written for the queue interface.

I think we're quite limited by our current architecture for setting up Gitea - I'd suggest a more DI approach but that's a massive PR.

Reply:
I think it is ok for settings to rely on queue because queue only relies on log (settings also relies on log), so circular dependency is not possible right now. I agree that DI should be done in future, but out of scope of this PR.